e2e/testcases/ssh_known_hosts_test.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/reconcilermanager/controllers"
+	"kpt.dev/configsync/pkg/status"
+)
+
+func TestSSHKnownHost(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.SyncSource, ntopts.SkipNonLocalGitProvider, ntopts.NamespaceRepo(backendNamespace, configsync.RepoSyncName))
+	var err error
+	rootSecret := &corev1.Secret{}
+	repoSecret := &corev1.Secret{}
+	nt.T.Cleanup(func() {
+		nt.MustMergePatch(rootSecret, secretDataDeletePatch(controllers.KnownHostsKey))
+		nt.MustMergePatch(repoSecret, secretDataDeletePatch(controllers.KnownHostsKey))
+		nomostest.DeletePodByLabel(nt, "app", reconcilermanager.ManagerName, true)
+	})
+
+	// get known host key value
+	knownHostValue, err := nomostest.GetKnownHosts(nt)
+	if err != nil {
+		nt.T.Fatalf("generating kh: %s", knownHostValue)
+	}
+
+	// create secret with known_hosts key
+	err = nt.Validate(nomostest.RootAuthSecretName, configsync.ControllerNamespace, rootSecret)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	err = nt.Watcher.WatchObject(kinds.Deployment(),
+		core.RootReconcilerPrefix, configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(reconcilermanager.GitSync, controllers.GitSyncKnownHosts, "false"),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	// apply known host key and validate
+	nt.MustMergePatch(rootSecret, secretDataPatch(controllers.KnownHostsKey, knownHostValue))
+
+	err = nt.Watcher.WatchObject(kinds.Deployment(),
+		core.RootReconcilerPrefix, configsync.ControllerNamespace,
+		[]testpredicates.Predicate{
+			testpredicates.DeploymentHasEnvVar(reconcilermanager.GitSync, controllers.GitSyncKnownHosts, "true"),
+		},
+	)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	// repo sync test
```

Comment:
This test currently covers the important assertions, but spreads them out across RootSync and RepoSync. I think it would be ideal to get full assertion coverage for both RootSync and RepoSync. For this it may be preferable to split this out into two tests - one for RootSync and one for RepoSync.

Reply:
the test has been split for rootsync and reposync and both will cover the full assertion